File: python/x402/src/bankofai/x402/signers/client/tron_signer.py
"""
TronClientSigner - TRON client signer implementation

Accepts any wallet object that exposes the agent-wallet BaseWallet interface
(get_address, sign_message, sign_typed_data, sign_transaction).
The signer is agnostic about how the wallet was created (private key, hosted, etc.).
"""

import json as json_module
import logging
from typing import Any

from bankofai.x402.abi import ERC20_ABI
from bankofai.x402.config import NetworkConfig
from bankofai.x402.exceptions import InsufficientAllowanceError, SignatureCreationError
from bankofai.x402.signers.client.base import ClientSigner

logger = logging.getLogger(__name__)


class TronClientSigner(ClientSigner):
    """TRON client signer implementation"""

    def __init__(self, wallet: Any) -> None:
        """Create signer from a wallet.

        Prefer the async factory ``create()`` or ``from_private_key()``.

        Args:
            wallet: Any object implementing the BaseWallet interface
                    (get_address, sign_message, sign_typed_data, sign_transaction).
        """
        self._wallet = wallet
        self._address: str | None = None
        self._async_tron_clients: dict[str, Any] = {}
        logger.debug("TronClientSigner initialized")

    @classmethod
    async def create(cls) -> "TronClientSigner":
        """Async factory: resolve active agent wallet and create signer."""
        from agent_wallet import resolve_wallet_provider

        provider = resolve_wallet_provider(network="tron")
        wallet = await provider.get_active_wallet()
        signer = cls(wallet)
        signer.set_address(await wallet.get_address())
        return signer

    @classmethod
    async def from_private_key(cls, private_key: str) -> "TronClientSigner":
        """Create signer from a raw private-key hex string (backward-compat).

        Uses agent-wallet's ``create_wallet_provider`` internally.
        """
        from agent_wallet import PrivateKeyProviderOptions, create_wallet_provider

        provider = create_wallet_provider(
            PrivateKeyProviderOptions(private_key=private_key, network="tron")
        )
        wallet = await provider.get_active_wallet()
        signer = cls(wallet)
        signer.set_address(await wallet.get_address())
        return signer

    @staticmethod
    def _extract_tron_signature(result: str) -> str:
        """Extract signature hex from agent-wallet's sign_transaction result.

        agent-wallet's TronWallet returns a JSON string with the full signed
        transaction.  We extract the first signature entry.
        """
        if isinstance(result, str) and result.strip().startswith("{"):
            signed = json_module.loads(result)
            sigs = signed.get("signature", [])
            if not sigs:
                raise ValueError("Wallet returned signed tx without signature")
            result = sigs[0]
        return result

    def _ensure_async_tron_client(self, network: str) -> Any:
        """Lazy initialize async tron_client for the given network.

        Args:
            network: Network identifier (e.g. 'tron:nile', 'tron:mainnet').

        Returns:
            tronpy.AsyncTron instance or None
        """
        if network not in self._async_tron_clients:
            try:
                from bankofai.x402.utils.tron_client import create_async_tron_client

                self._async_tron_clients[network] = create_async_tron_client(network)
            except ImportError:
                return None
        return self._async_tron_clients[network]

    def get_address(self) -> str:
        if not self._address:
            raise ValueError("Signer address has not been initialized")
        return self._address

    def set_address(self, address: str) -> None:
        self._address = address

    async def sign_message(self, message: bytes) -> str:
        """Sign raw message using ECDSA"""
        try:
            return await self._wallet.sign_message(message)
        except Exception as e:
            raise SignatureCreationError(f"Failed to sign message: {e}")

    async def sign_typed_data(
        self,
        domain: dict[str, Any],
        types: dict[str, Any],
        message: dict[str, Any],
        primary_type: str,
    ) -> str:
        """Sign EIP-712 typed data (Pure Passthrough)."""
        logger.info(
            f"Signing EIP-712 typed data: domain={domain.get('name')}, primaryType={primary_type}"
        )
        try:
            typed_data = {
                "types": types,
                "primaryType": primary_type,
                "domain": domain,
                "message": message,
            }

            # Log domain and message in same format as TypeScript client
            import json as json_module

            # Convert bytes to hex for logging
            message_for_log = dict(message)
            if "meta" in message_for_log and "paymentId" in message_for_log["meta"]:
                pid = message_for_log["meta"]["paymentId"]
                if isinstance(pid, bytes):
                    message_for_log["meta"] = dict(message_for_log["meta"])
                    message_for_log["meta"]["paymentId"] = "0x" + pid.hex()

            logger.info(f"[SIGN] Domain: {json_module.dumps(domain)}")
            logger.info(f"[SIGN] Message: {json_module.dumps(message_for_log)}")

            signature = await self._wallet.sign_typed_data(typed_data)
            logger.info(f"[SIGN] Signature: 0x{signature}")
            return signature
        except Exception as e:
            raise SignatureCreationError(f"Failed to sign typed data: {e}")

    async def check_balance(
        self,
        token: str,
        network: str,
        address: str | None = None,
    ) -> int:
        """Check TRC20 token balance"""
        client = self._ensure_async_tron_client(network)
        if client is None:
            logger.warning("AsyncTron client not available, returning 0 balance")
            return 0

        target_address = address or self.get_address()
        try:
            contract = await client.get_contract(token)
            contract.abi = ERC20_ABI
            balance = await contract.functions.balanceOf(target_address)
            balance_int = int(balance)
            from bankofai.x402.tokens import TokenRegistry

            token_info = TokenRegistry.find_by_address(network, token)
            decimals = token_info.decimals if token_info else 6
            symbol = token_info.symbol if token_info else token[:8]
            human = balance_int / (10**decimals)
            logger.info(
                f"Token balance for {target_address}: {human:.6f} {symbol} "
                f"(raw={balance_int}, token={token}, network={network})"
            )
            return balance_int
        except Exception as e:
            logger.error(f"Failed to check balance for {target_address}: {e}")
            return 0

    async def check_allowance(
        self,
        token: str,
        amount: int,
        network: str,
    ) -> int:
        """Check token allowance on TRON"""
        spender = self._get_spender_address(network)
        address = self.get_address()
        logger.info(
            "Checking allowance: token=%s, owner=%s, spender=%s, network=%s",
            token,
            address,
            spender,
            network,
        )
        if not spender or spender == "T0000000000000000000000000000000":
            logger.warning(
                f"Invalid spender address for network {network}, skipping allowance check"
            )
            return 0

        client = self._ensure_async_tron_client(network)
        if client is None:
            logger.warning("AsyncTron client not available, returning 0 allowance")
            return 0

        try:
            contract = await client.get_contract(token)
            contract.abi = ERC20_ABI
            allowance = await contract.functions.allowance(
                address,
                spender,
            )
            allowance_int = int(allowance)
            logger.info(f"Current allowance: {allowance_int}")
            return allowance_int
        except Exception as e:
            logger.error(f"Failed to check allowance: {e}")
            return 0

    async def ensure_allowance(
        self,
        token: str,
        amount: int,
        network: str,
        mode: str = "auto",
    ) -> bool:
        """Ensure sufficient allowance"""
        logger.info(
            f"Ensuring allowance: token={token}, amount={amount}, network={network}, mode={mode}"
        )
        if mode == "skip":
            logger.info("Skipping allowance check (mode=skip)")
            return True

        current = await self.check_allowance(token, amount, network)
        if current >= amount:
            logger.info(f"Sufficient allowance already exists: {current} >= {amount}")
            return True

        if mode == "interactive":
            raise NotImplementedError("Interactive approval not implemented")

        logger.info(f"Insufficient allowance ({current} < {amount}), requesting approval...")
        client = self._ensure_async_tron_client(network)
        if client is None:
            raise InsufficientAllowanceError("AsyncTron client required for approval")

        try:
            spender = self._get_spender_address(network)
            # Use maxUint160 (2^160 - 1) to avoid repeated approvals
            max_uint160 = (2**160) - 1
            logger.info(f"Approving spender={spender} for amount={max_uint160} (maxUint160)")
            contract = await client.get_contract(token)
            contract.abi = ERC20_ABI
            # AsyncTron: contract.functions.approve() returns a coroutine, need to await it first
            txn_builder = await contract.functions.approve(spender, max_uint160)
            owner_address = self.get_address()
            txn_builder = txn_builder.with_owner(owner_address).fee_limit(100_000_000)
            txn = await txn_builder.build()
            # Sign the transaction via wallet
            txn_dict = txn.to_json()
            raw_result = await self._wallet.sign_transaction(txn_dict)
            sig_hex = self._extract_tron_signature(raw_result)
            txn._signature = [sig_hex]
            logger.info("Broadcasting approval transaction...")
            result = await txn.broadcast()
            result = await result.wait()
            # Check receipt.result for success (TRON returns "SUCCESS" in receipt)
            receipt = result.get("receipt", {})
            receipt_result = receipt.get("result", "")
            success = receipt_result == "SUCCESS"
            if success:
                logger.info(f"Approval successful: txid={result.get('id')}")
            else:
                logger.warning(f"Approval failed: {result}")
            return success
        except Exception as e:
            raise InsufficientAllowanceError(f"Approval transaction failed: {e}") from e

    def _get_spender_address(self, network: str) -> str:
        """Get payment permit contract address (spender)"""
        return NetworkConfig.get_payment_permit_address(network)

File: typescript/packages/x402/src/signers/facilitator/tronSigner.ts
/**
 * TronFacilitatorSigner — TRON facilitator signer (TIP-712 verify + contract write).
 *
 * Mirrors Python `bankofai.x402.signers.facilitator.tron_signer.TronFacilitatorSigner`.
 *
 * Uses:
 * - `tronweb` v6 for transaction building, signing wrap, broadcasting, receipt polling
 * - `viem` for cross-chain EIP-712 signature recovery (TRON TIP-712 is structurally
 *   identical to EIP-712 once addresses are normalized to 0x hex)
 * - `@bankofai/agent-wallet` for the signing wallet abstraction
 */

import { recoverTypedDataAddress } from 'viem';
import type { TypedDataDomain as ViemTypedDataDomain } from 'viem';
import { resolveWalletProvider } from '@bankofai/agent-wallet';
import { TronWeb as TronWebClass } from 'tronweb';

import {
  FacilitatorSigner,
  type TypedDataDomain,
  type TypedDataTypes,
  type TransactionReceipt,
} from './base.js';
import type { TronWeb } from '../types.js';
import { toEvmHex } from '../../address.js';
import { getTronRpcUrl } from '../../config.js';
import type { AgentWallet } from '../signer.js';

/**
 * Buffer-like check that works in both Node and bundled environments without
 * requiring `Buffer` to be in the global scope.
 */
function hexToBytes(hex: string): Uint8Array {
  const clean = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (clean.length % 2 !== 0) {
    throw new Error(`Invalid hex string length: ${clean.length}`);
  }
  const out = new Uint8Array(clean.length / 2);
  for (let i = 0; i < out.length; i++) {
    out[i] = parseInt(clean.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

/**
 * TRON facilitator signer.
 *
 * The wallet is responsible for signing transactions; the signer wraps tronweb
 * for building and broadcasting them. For verify, we use viem's typed-data
 * recovery directly — it does not need the wallet.
 */
export class TronFacilitatorSigner extends FacilitatorSigner {
  private wallet: AgentWallet;
  private address: string = ''; // Base58 format
  private tronWebInstances: Map<string, TronWeb> = new Map();
  /**
   * Raw private key (hex, no `0x`) read from `TRON_FACILITATOR_PRIVATE_KEY`
   * or `TRON_PRIVATE_KEY`. When present, signs transactions directly via
   * `tronweb.trx.sign(tx, privateKey)` — mirrors Python's tronpy direct-sign
   * path that bypasses agent-wallet (because agent-wallet's TRON sign_transaction
   * round-trip can drift the tx's raw_data and invalidate the txID/signature).
   */
  private privateKey: string | null = null;
  /**
   * TRON permission id used when building transactions.
   * Default `2` = active permission (required for multi-sig facilitator
   * accounts with owner threshold > 1). Override via `TRON_PERMISSION_ID` env.
   */
  private permissionId: number = 2;

  constructor(wallet: AgentWallet) {
    super();
    this.wallet = wallet;
    if (typeof process !== 'undefined') {
      const pk =
        process.env?.TRON_FACILITATOR_PRIVATE_KEY ??
        process.env?.TRON_PRIVATE_KEY ??
        null;
      if (pk) {
        this.privateKey = pk.startsWith('0x') ? pk.slice(2) : pk;
      }
      const permId = process.env?.TRON_PERMISSION_ID;
      if (permId) {
        const parsed = parseInt(permId, 10);
        if (!isNaN(parsed)) this.permissionId = parsed;
      }
    }
  }

  /** Async factory: resolve active agent wallet and create signer. */
  static async create(): Promise<TronFacilitatorSigner> {
    const provider = resolveWalletProvider({ network: 'tron' });
    const wallet = await provider.getActiveWallet();
    const signer = new TronFacilitatorSigner(wallet as unknown as AgentWallet);
    signer.setAddress(await wallet.getAddress());
    return signer;
  }

  setAddress(address: string): void {
    this.address = address;
  }

  getAddress(): string {
    if (!this.address) {
      throw new Error('TronFacilitatorSigner address has not been initialized');
    }
    return this.address;
  }

  /**
   * Extract signature hex from agent-wallet's signTransaction result.
   * agent-wallet TronWallet returns a JSON string with the full signed tx.
   */
  private extractTronSignature(result: string): string {
    if (typeof result === 'string' && result.trim().startsWith('{')) {
      const signed = JSON.parse(result);
      const sigs = signed.signature;
      if (!Array.isArray(sigs) || sigs.length === 0) {
        throw new Error('Wallet returned signed tx without signature');
      }
      return sigs[0];
    }
    return result;
  }

  /** Get or create a TronWeb instance for the given network. */
  private getTronWeb(network: string): TronWeb {
    const host = getTronRpcUrl(network);
    if (!host) {
      throw new Error(`No TRON RPC URL configured for network: ${network}`);
    }
    let tw = this.tronWebInstances.get(host);
    if (!tw) {
      const apiKey =
        typeof process !== 'undefined' ? process.env?.TRON_GRID_API_KEY : undefined;
      const headers = apiKey ? { 'TRON-PRO-API-KEY': apiKey } : undefined;
      // Dummy private key — facilitator signing routes via wallet.signTransaction,
      // not via TronWeb's defaultPrivateKey.
      const dummyKey =
        '0000000000000000000000000000000000000000000000000000000000000001';
      tw = new TronWebClass({
        fullHost: host,
        privateKey: dummyKey,
        headers,
      }) as unknown as TronWeb;
      this.tronWebInstances.set(host, tw);
    }
    return tw;
  }

  /**
   * Verify EIP-712 / TIP-712 signature.
   *
   * Strategy: convert all TRON Base58 addresses in domain.verifyingContract
   * and the message to 0x EVM hex, then use viem's `recoverTypedDataAddress`
   * to recover the signer. Compare against the expected address (also
   * converted to hex).
   *
   * Domain types are NOT included in the `types` arg here — viem injects
   * `EIP712Domain` automatically.
   */
  async verifyTypedData(
    address: string,
    domain: TypedDataDomain,
    types: TypedDataTypes,
    message: Record<string, unknown>,
    signature: string,
    primaryType: string,
  ): Promise<boolean> {
    try {
      // viem expects hex domain and hex addresses; convert anything Base58.
      const viemDomain: ViemTypedDataDomain = {
        name: domain.name,
        version: domain.version,
        chainId: domain.chainId,
        verifyingContract: domain.verifyingContract
          ? (toEvmHex(domain.verifyingContract) as `0x${string}`)
          : undefined,
        ...(domain.salt ? { salt: domain.salt as `0x${string}` } : {}),
      };

      // Strip an `EIP712Domain` entry if a caller mistakenly added it; viem
      // injects this from the domain object.
      const typesNoDomain: TypedDataTypes = { ...types };
      delete (typesNoDomain as Record<string, unknown>)['EIP712Domain'];

      const sigBytes = signature.startsWith('0x')
        ? (signature as `0x${string}`)
        : (`0x${signature}` as `0x${string}`);

      const recovered = await recoverTypedDataAddress({
        domain: viemDomain,
        types: typesNoDomain as Record<string, ReadonlyArray<{ name: string; type: string }>>,
        primaryType,
        message,
        signature: sigBytes,
      });

      const expectedHex = toEvmHex(address).toLowerCase();
      return recovered.toLowerCase() === expectedHex;
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error('[TronFacilitatorSigner] verifyTypedData failed:', err);
      return false;
    }
  }

  /**
   * Write contract on TRON: build via tronweb's `triggerSmartContract`, sign
   * via the wallet, broadcast.
   *
   * The function selector must match the on-chain ABI exactly, including
   * tuple bracket types. Example for PaymentPermit:
   *
   *   permitTransferFrom(((uint8,bytes16,uint256,uint256,uint256),
   *                       address,address,(address,uint256,address),(address,uint256)),
   *                      address,bytes)
   *
   * @param contractAddress Base58 (T...) or hex (41...) contract address.
   * @param _abi Unused for TRON (we use function selector directly); kept for interface parity.
   * @param method Method signature string with full type bracket layout (see note above).
   * @param args Parameter array shaped per tronweb's `triggerSmartContract` schema:
   *             `[{ type: 'address', value: '0x...' }, { type: 'tuple(...)', value: [...] }, ...]`.
   */
  async writeContract(
    contractAddress: string,
    _abi: string | unknown[],
    method: string,
    args: unknown[],
    network: string,
  ): Promise<string | null> {
    try {
      const tw = this.getTronWeb(network);

      const tx = await tw.transactionBuilder.triggerSmartContract(
        contractAddress,
        method,
        {
          feeLimit: 1_000_000_000, // 1000 TRX cap
          callValue: 0,
          // Active permission (id=2) required for multi-sig facilitator
          // accounts whose owner threshold is > 1.
          permissionId: this.permissionId,
        },
        args,
        this.getAddress(),
      );

      if (!tx.result?.result) {
        // eslint-disable-next-line no-console
        console.error(
          `[TronFacilitatorSigner] triggerSmartContract failed: ${JSON.stringify(tx)}`,
        );
        return null;
      }

      // Sign: prefer direct private-key signing via tronweb (matches Python's
      // tronpy direct-sign path). agent-wallet's TRON signTransaction can
      // round-trip the raw_data and produce an invalid signature.
      let signedTx: unknown;
      if (this.privateKey) {
        signedTx = await (tw.trx as unknown as {
          sign: (txObj: unknown, pk: string) => Promise<unknown>;
        }).sign(tx.transaction, this.privateKey);
      } else {
        const rawResult = await this.wallet.signTransaction(
          tx.transaction as Record<string, unknown>,
        );
        const sigHex = this.extractTronSignature(rawResult);
        signedTx = {
          ...(tx.transaction as Record<string, unknown>),
          signature: [sigHex],
        };
      }

      // Broadcast
      const broadcast = await tw.trx.sendRawTransaction(signedTx);
      if (!broadcast.result) {
        // eslint-disable-next-line no-console
        console.error(
          `[TronFacilitatorSigner] sendRawTransaction failed: ${JSON.stringify(broadcast)}`,
        );
        return null;
      }
      return broadcast.txid;
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error('[TronFacilitatorSigner] writeContract error:', err);
      return null;
    }
  }

  /**
   * Check TRC20 balance via `triggerConstantContract` (read-only call).
   */
  async checkBalance(
    token: string,
    network: string,
    address?: string,
  ): Promise<bigint> {
    try {
      const targetAddress = address ?? this.getAddress();
      const ownerHex = toEvmHex(targetAddress);

      const tw = this.getTronWeb(network);
      const result = await tw.transactionBuilder.triggerConstantContract(
        token,
        'balanceOf(address)',
        {},
        [{ type: 'address', value: ownerHex }],
        this.getAddress(),
      );

      if (result.result?.result && result.constant_result?.length) {
        return BigInt('0x' + result.constant_result[0]);
      }
      // Suppress unused vars from helper imports
      void hexToBytes;
    } catch (err) {
      // eslint-disable-next-line no-console
      console.error(`[TronFacilitatorSigner] checkBalance failed for ${token}:`, err);
    }
    return BigInt(0);
  }

  /**
   * Poll tronpy `getTransactionInfo` until the tx has a blockNumber.
   *
   * Returns `{status: 'confirmed' | 'failed'}` based on `receipt.result`.
   */
  async waitForTransactionReceipt(
    txHash: string,
    network: string,
    timeoutMs: number = 120_000,
  ): Promise<TransactionReceipt> {
    const tw = this.getTronWeb(network);
    const deadline = Date.now() + timeoutMs;

    while (Date.now() < deadline) {
      try {
        const info = await tw.trx.getTransactionInfo(txHash);
        if (info && info.blockNumber) {
          const success = info.receipt?.result === 'SUCCESS';
          return {
            hash: txHash,
            blockNumber: String(info.blockNumber),
            status: success ? 'confirmed' : 'failed',
          };
        }
      } catch {
        // Not yet propagated to the queried node — keep polling.
      }
      await new Promise((resolve) => setTimeout(resolve, 3_000));
    }

    throw new Error(
      `Transaction ${txHash} not confirmed within ${Math.round(timeoutMs / 1000)}s`,
    );
  }
}
